### ISSUE TITLE
Support GitHub Merge Queue

### ISSUE THREAD

**Issue by wrslatz** [palantir/bulldozer#297]
GitHub [announced](https://github.blog/changelog/2021-10-27-pull-request-merge-queue-limited-beta/) a new merge queue capability that is currently in limited beta. It allows an additional flow path for merging pull requests by adding them to a merge queue (handled by GitHub Actions behind the scenes). This feature is still relatively undocumented, so we can't take action on this yet, but adding this feature to Bulldozer would help Bulldozer integrate well with this new functionality from GitHub.

Additional resources:
- Docs on [using a merge queue](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/using-a-merge-queue) and [adding a pull request to the merge queue](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/adding-a-pull-request-to-the-merge-queue)
-  [GraphQL API](https://docs.github.com/en/graphql/reference/mutations#enablepullrequestautomerge) for enabling the merge queue for a PR
- https://github.com/github/feedback/discussions/8306, where GitHub says they currently do not plan to automate adding a PR to the merge queue at the platform level

Next steps before implementation could be planned:
- Review API documentation around auto-merge
- Track the corresponding GitHub roadmap issues for full merge queue support (https://github.com/github/roadmap/issues/370, https://github.com/github/roadmap/issues/455)
- Consider implementation
    - This seems separate from the existing merge methods

Additional features to consider once this one is implemented:
- Matching to move to merge queue (trigger, ignore, labels, etc.)
- Matching to jump ahead of the merge queue (trigger, ignore, labels, etc.)

**Comment by asvoboda** [comment]
Hey @wrslatz,

Thanks for filing this. I think the idea of integrating with the merge queue makes a lot of sense. I think we'll want to wait until the feature is out of public beta to let the API stabilize.

On your points of discussion:

* Should this feature be implemented as a new merge-method, or a separate construct? We may want a separate option, since the merge queue takes as input a merge method and pull request to perform the merge itself.

**Comment by wrslatz** [comment]
Looks like the merge queue is now exposed via API in this [latest update](https://github.blog/changelog/2023-04-19-pull-request-merge-queue-public-beta-api-support-and-recent-fixes/)! One step closer to GA and adding support for this feature

**Comment by wrslatz** [comment]
https://github.com/palantir/policy-bot/pull/554 might be good as reference

**Comment by RajatPorwal5** [comment]
Is this something currently being worked on? There are couple of folks looking for this feature.

**Comment by bluekeyes** [comment]
No, we do not use merge queues and are not currently working on this, but are open to PRs if anyone else would like to try implementing merge queue support.

**Comment by reegnz** [comment]
I started looking into this as my organization wants to adopt merge queues and we're using bulldozer.

Implementing merge-queue support will need GraphQL because the Merge Queue functionality is not exposed through the REST API, only GraphQL: https://docs.github.com/en/graphql/reference/mutations?versionId=free-pro-team%40latest&page=objects#enqueuepullrequest

Apparently the current github client used in palantir/go-githubapp does provide a way to use v4clients too, so I'll give it a go and try to implement.
